Add Awair Local API support

zachberger opened the pull request:
## Proposed change
Add support for Awair's (beta) Local API. This API supports more frequent polling and will work when there is no Internet connection. 

Previous attempt in #39538 but that was introducing an entirely new integration. #39538#issuecomment-693547702 suggested extending the current awair component to add an option for the local API, which this new PR does. #39538 also suggested mDNS discovery, that is not in this PR, but could be explored and submitted in a future PR.

Still to be done:
- [ ] More tests need to be written
- [x] more thoughtful unique IDs for local API
- [x] more thoughtful strings for ConfigFlow

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: N/A
- This PR is related to issue: #39538
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/23588

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works. TODO

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

zachberger committed:
Add Awair Local API support

ahayworth reviewed (commented):
🎉 🎉 

Yes, I think this is the right approach. I keep thinking "gosh, we should really come back to the local api" but I've had approximately zero time/drive to do so. Thank you so much for taking up the charge!

I will reserve thorough review until it's marked as ready, but I do think the overall direction at the moment is correct.

Regarding mDNS discovery: yes, that can come later. In the future, I think we should have a flow that prefers local by default (and does mDNS discovery for you) but accepts an API token to still support the oldest generation devices. But like you proposed: we'll do that next. A list of hosts is a good next step and we need momentum to keep going. This gives us that momentum!

Thank you! Excited to review it when it's done! 😄

zachberger committed:
Fix awair ConfigFlow to pass existing tests for Cloud API

zachberger committed:
When connecting locally allow only a single host

This allows us to use the MAC address as the unique identifier, and
allows each Awair to be discovered separately.

zachberger committed:
Add Awair local device discovery MVP

zachberger commented:
Slowly making some progress on this. Its summer so not too much free time. 

* I was able to get discover working fairly simply in [5d4efb2](https://github.com/home-assistant/core/pull/75535/commits/5d4efb2bc986ff7708d05c1c305c3c670d7677c7)
* Started working on documentation in https://github.com/home-assistant/home-assistant.io/pull/23588

zachberger commented on homeassistant/components/awair/manifest.json:
```diff
@@ -6,5 +6,12 @@
   "iot_class": "cloud_polling",
-  "loggers": ["python_awair"]
+  "loggers": ["python_awair"],
+  "after_dependencies": ["zeroconf"],
```
I'm not certain this is needed. Some components with zeroconf discovery declare it, others do not.

balloob replied:
you don't need it unless you import from it. Hassfest will tell you.

zachberger commented on homeassistant/components/awair/manifest.json:
```diff
@@ -6,5 +6,12 @@
   "codeowners": ["@ahayworth", "@danielsjf"],
   "config_flow": true,
   "iot_class": "cloud_polling",
```
I'm looking for guidance on what value to set this as the integration will support both cloud polling and local polling after this change.  Ideally on the integrations page it will show the cloud badge when configured to use the cloud api and not show that badge when using the local API. As far as I can tell that badge is driven by the value set in the manifest.

balloob replied:
local

zachberger committed:
Fixup some Awair typings

balloob commented on homeassistant/components/awair/__init__.py:
```diff
@@ -74,9 +104,27 @@ async def _async_update_data(self) -> Any | None:
+        async with timeout(API_TIMEOUT):
+            try:
+                LOGGER.debug("Fetching devices")
+                devices = await self._awair.devices()
+                results = await gather(
```
Why does the local awair update coordinator need to refresh devices?

I would assume each config entry to be bound to a single local Awair device, and we should have a data update coordinator per device. That way if entities of 1 device are disabled, we stop polling.

zachberger replied:
As mentioned below, calling `devices()` is canonical way to interact with `python_awair`, that being said I think this could be simplified to something like:

```py
if self._device is None:
    LOGGER.debug("Fetching devices")
    devices = await self._awair.devices()
    self._device = devices[0]
result = self._fetch_air_data(self._device) 
return {result.device.uuid: result}
```

balloob commented on homeassistant/components/awair/__init__.py:
```diff
@@ -74,9 +104,27 @@ async def _async_update_data(self) -> Any | None:
+                    *(self._fetch_air_data(device) for device in devices)
+                )
+                return {result.device.uuid: result for result in results}
+            except Exception as err:
```
Be specific on exceptions that you expect.

balloob commented on homeassistant/components/awair/config_flow.py:
```diff
@@ -21,20 +23,79 @@ class AwairFlowHandler(ConfigFlow, domain=DOMAIN):
     ) -> FlowResult:
         """Handle a flow initialized by the user."""
+
+        if user_input is not None:
+            if CONF_ACCESS_TOKEN in user_input:
+                return await self.async_step_cloud(user_input)
+            if CONF_HOST in user_input:
+                return await self.async_step_local(user_input)
+
```
This code is never reached. A menu will automatically redirect to the next step based on the picked option.

```suggestion
```

balloob commented on homeassistant/components/awair/config_flow.py:
```diff
@@ -79,7 +166,27 @@ async def async_step_reauth_confirm(
+        awair = AwairLocal(session=session, device_addrs=[device_address])
+
+        try:
+            devices = await awair.devices()
+            if not devices:
+                return (None, "no_devices")
```
Why do we need to call `devices` if we are targeting a single device via a local connection ?

zachberger replied:
This seems to be the canonical way to interact with `python_awair`. The library is built around working with multiple devices. That being said, I was able to change and simplify this to just use a single device

balloob commented on homeassistant/components/awair/translations/pt-BR.json:
```diff
@@ -24,7 +24,7 @@
                 "description": "Insira novamente seu token de acesso de desenvolvedor Awair."
             },
-            "user": {
+            "cloud": {
```
No need to update `translations` folder. The source of truth is lokalise.

For local testing, run `python3 -m script.translations develop` to generate dev translations.

balloob commented on homeassistant/components/awair/config_flow.py:
```diff
@@ -21,20 +23,79 @@ class AwairFlowHandler(ConfigFlow, domain=DOMAIN):
                 self._abort_if_unique_id_configured()
 
-                title = f"{user.email} ({user.user_id})"
+                title = f"Awair Cloud {user.email} ({user.user_id})"
```
This should just be the user.email. Users don't care about their ID or that it's cloud. 
```suggestion
                title = user.email
```

balloob commented on homeassistant/components/awair/config_flow.py:
```diff
@@ -21,20 +23,79 @@ class AwairFlowHandler(ConfigFlow, domain=DOMAIN):
+    ) -> FlowResult:
+        """Confirm discovery."""
+        if user_input is not None:
+            title = f"{self._device.model} ({self._device.device_addr})"
```
The address can change if the IP address changes. So let's not put it in the name.

Can we fetch the name of the Awair device if the user has given it one?

zachberger replied:
Name isn't returned in the local API. Here is a dump of whats available locally:

```json
{
  "device_uuid":"awair-element_12345",
  "wifi_mac":"REDACTED",
  "ssid":"Internet of Shit",
  "ip":"10.0.5.188",
  "netmask":"255.255.255.0",
  "gateway":"none",
  "fw_version":"1.2.8",
  "timezone":"America/Los_Angeles",
  "display":"score",
  "led":{
    "mode":"auto",
    "brightness":179
  },
  "voc_feature_set":34
}
```

zachberger replied:
I've removed the host for now. I think there should be some unique identifier shown so that the user knows which device they're setting up, but its not clear what the right choice is here.

zachberger replied:
I think I will do `f"{self._device.model} ({self._device.device_id})"`

`device_id` is the last 5 digits on `device_uuid`

balloob commented on homeassistant/components/awair/translations/en.json:
```diff
@@ -1,14 +1,17 @@
 {
```
All these changes are all going to be lost. Update `strings.json` instead.

balloob commented on homeassistant/components/awair/translations/en.json:
```diff
@@ -1,14 +1,17 @@
     "config": {
         "abort": {
-            "already_configured": "Account is already configured",
+            "already_configured": "Device or Awair account is already configured",
```
There are 2 already configured error messages. Don't merge them. One for account, one for device.

zachberger replied:
How can `_abort_if_unique_id_configured()` know which string to use?

balloob replied:
Let's allow to override it. 

In `config_entries.py`, change

```python
    @callback
    def _abort_if_unique_id_configured(
        self,
        updates: dict[str, Any] | None = None,
        reload_on_update: bool = True,
    ) -> None:
```

to

```python
    @callback
    def _abort_if_unique_id_configured(
        self,
        updates: dict[str, Any] | None = None,
        reload_on_update: bool = True,
        *,
        error: str = "already_configured"
    ) -> None:
```

balloob commented on homeassistant/components/awair/translations/en.json:
```diff
@@ -24,12 +27,28 @@
+                "data": {
+                    "host": "Hostname or IP address of the device"
+                },
+                "description": "Awair Local API must be enabled following these steps: https://support.getawair.com/hc/en-us/articles/360049221014-Awair-Element-Local-API-Feature"
```
Inject URL via `description_placeholders`.

balloob commented on homeassistant/components/awair/translations/en.json:
```diff
@@ -24,12 +27,28 @@
+                    "cloud": "Cloud API",
+                    "local": "Local API"
+                  },
+                "description": "Awair provides two APIs, a Cloud API with data summarized over 5 minutes, and a beta Local API which can be polled for real time data."
```
We should steer everyone to local. It's local and faster. Cloud should only be used for devices not on the local LAN.

So we should make the description represent that.

Pick local for the best experience. Only use cloud if the device is not connected to the same network as Home Assistant.

balloob commented on homeassistant/components/awair/translations/en.json:
```diff
@@ -24,12 +27,28 @@
+            },
+            "user": {
+                "menu_options": {
+                    "cloud": "Cloud API",
+                    "local": "Local API"
```
Drop word "API". Not user friendly. 

- Connect locally (preferred)
- Connect via the cloud

balloob commented on homeassistant/components/awair/config_flow.py:
```diff
@@ -21,20 +23,79 @@ class AwairFlowHandler(ConfigFlow, domain=DOMAIN):
+                }
+            )
+        else:
+            return self.async_abort(reason="unreachable")
```
Can we differentiate between the device not being reachable versus the device is an Awair device but the local API is disabled? Because in the latter case we shouldn't abort, but instead guide the user on how to enable their local API.

balloob replied:
This one is still open. Let's address this in a future PR. My Awair Element is coming in on Friday so I can help improve this experience.

zachberger replied:
Sure, makes sense improve it in the future. FWIW, the API is exposed on `:80` and the device does not listen on this port until the API is enabled. Perhaps ICMP can be used to check recachability?

balloob replied:
Does the device only broadcast mDNS when the API is enabled?

zachberger replied:
That is correct, mDNS is only broadcast when the API is enabled. That being said, hass receiving a mDNS packet doesn't necessarily mean that hass will be able to connect back to the awair device. This happened in my environment initially where I allowed mDNS broadcasts across VLAN, but firewall was preventing the tcp connection.

zachberger committed:
Refine Awair Local integration

* Update title of cloud and local integrations
* Remove un-needed code
* Catch more specific errors in AwairLocalDataUpdateCoordinator
* Don't refresh devices on every call to _async_update_data
* Improve typings
* Set proper value for iot_class in manifest.json

zachberger committed:
Add ConfigFlow test for Awair local devices

zachberger committed:
Allow override of error in _abort_if_unique_id_configured

balloob commented on .gitignore:
```diff
@@ -132,3 +132,6 @@ tmp_cache
 # python-language-server / Rope
 .ropeproject
+
+# Generated test files
```
Please remove this.

zachberger replied:
Is there a reason to not have them ignored?

balloob replied:
I don't know what it is.

zachberger replied:
When I run tox it is creating a folder `homeassistant-2022.9.0.dev0` with thousands of unignored files. It looks like its building a tar file during the build process, and this folder is the staging area. Anyways, when the thousands of files are dropped on the FS it crashes VS Code's git integration. I've found adding `*.dev0/` this folder prevents the crash.

balloob replied:
Ah. So `dev0` here is just the version number of dev, can't be used as a matcher. I don't run `tox` myself but always use `pytest tests/components/awair`. Let me check on this.

zachberger replied:
Removed the change for now, I will see if there is a way I can exclude it locally

balloob replied:
Checked with some others and the conclusion was that no one actually uses Tox 😅 We're going to remove it and rewrite the dev docs https://github.com/home-assistant/core/pull/76582

balloob commented on homeassistant/components/awair/translations/bg.json:
```diff
@@ -13,7 +13,7 @@
                     "email": "Email"
                 }
             },
-            "user": {
```
Please remove all changes to `translations` folder from your PR to avoid merge conflicts.

zachberger replied:
Yes, still working through the unresolved comments

balloob commented on tests/components/awair/const.py:
```diff
@@ -2,14 +2,17 @@
+LOCAL_CONFIG = {CONF_HOST: "192.0.2.5"}
+CLOUD_UNIQUE_ID = "foo@bar.com"
+LOCAL_UNIQUE_ID = "00:B0:D0:63:C2:26"
+CLOUD_DEVICES_FIXTURE = json.loads(load_fixture("awair/cloud_devices.json"))
```
let's not load a fixture at the top level. Instead put it in a fixture so it's only loaded when a test that uses it is run. Scope=session makes sure it is only loaded once every test run:

https://github.com/home-assistant/core/blob/4a938ec33ee8df17e619fc64a519ef7c2ff856d8/tests/components/zwave_js/conftest.py#L226-L229

balloob commented on tests/components/awair/fixtures/local_devices.json:
```diff
@@ -0,0 +1,16 @@
+{
```
So annoying it doesn't have the name in it. We should reach out to Awair to request this.

zachberger replied:
Agreed, its unfortunate

balloob reviewed (approved):
Looks great! Just few tiny things and this would be good to go 🎉

zachberger committed:
Revert changes made in Awair translations folder

zachberger committed:
Update strings for awair local devices

zachberger committed:
Migrate Awair test fixtures to pytest.fixture

MartinHjelmare commented on tests/components/awair/test_config_flow.py:
```diff
@@ -1,99 +1,139 @@
+        "python_awair.AwairClient.query",
+        side_effect=[user, cloud_devices],
     ), patch(
         "homeassistant.components.awair.sensor.async_setup_entry",
```
We only need to patch the integration setup function on create entry results or if the config entry is reloaded, eg when re-authenticating the config entry.

MartinHjelmare commented on tests/components/awair/test_config_flow.py:
```diff
@@ -140,27 +183,65 @@ async def test_reauth_error(hass: HomeAssistant) -> None:
+        "python_awair.AwairClient.query",
+        side_effect=[user, cloud_devices],
     ), patch(
         "homeassistant.components.awair.sensor.async_setup_entry",
```
```suggestion
        "homeassistant.components.awair.async_setup_entry",
```

MartinHjelmare commented on tests/components/awair/test_config_flow.py:
```diff
@@ -140,27 +183,65 @@ async def test_reauth_error(hass: HomeAssistant) -> None:
+    """Test overall flow."""
+
+    with patch("python_awair.AwairClient.query", side_effect=[local_devices]), patch(
+        "homeassistant.components.awair.sensor.async_setup_entry",
```
```suggestion
        "homeassistant.components.awair.async_setup_entry",
```

zachberger committed:
Cleanup Awair test patching

* Only patch when needed
* Patch correct methods

zachberger commented:
I have more tests to add before this is done, but @ahayworth could you PTAL too, I'd appreciate your feedback on the state of the integration.

balloob commented on homeassistant/components/awair/__init__.py:
```diff
@@ -2,28 +2,42 @@
+
+    if CONF_HOST in config_entry.data:
+        coordinator = AwairLocalDataUpdateCoordinator(hass, config_entry, session)
+    elif CONF_ACCESS_TOKEN in config_entry.data:
```
I guess this can just be `else`
```suggestion
    else:
```

balloob commented on homeassistant/components/awair/__init__.py:
```diff
@@ -74,9 +104,33 @@ async def _async_update_data(self) -> Any | None:
+                    self._device = devices[0]
+                result = await self._fetch_air_data(self._device)
+                return {result.device.uuid: result}
+            except ClientConnectorError as err:
```
No need to catch these, data update coordinator already handles all aiohttp errors

zachberger committed:
Add first test for Awair zeroconf discovery.

zachberger committed:
Simplify Awair setup and local update coordinator.

* Don't catch exceptions caught elsewhere
* elif to else

balloob commented on homeassistant/components/awair/strings.json:
```diff
@@ -1,29 +1,49 @@
       },
+      "local": {
+        "data": {
+          "host": "Hostname or IP address of the device"
```
From what I recall, Awair mentions only IP addresses in their documentation. We should do the same:
```suggestion
          "host": "[%key:common::config_flow::data::ip%]"
```

balloob commented on homeassistant/components/awair/strings.json:
```diff
@@ -1,29 +1,49 @@
+          "cloud": "Connect via the cloud",
+          "local": "Connect locally (preferred)"
+        },
+        "description": "Pick local for the best experience. Only use cloud if the device is not connected to the same network as Home Assistant, of if you have a legacy device."
```
```suggestion
        "description": "Pick local for the best experience. Only use cloud if the device is not connected to the same network as Home Assistant, or if you have a legacy device."
```

zachberger committed:
Add tests for Awair local device setup when device is unreachable.

zachberger committed:
Fixup strings for Awair local devices

zachberger committed:
Remove Awair setup error handling for impossible error

zachberger committed:
Add test for when Awair throws error when connecting locally

zachberger committed:
Add test for updating data from local Awair

zachberger committed:
Fix docstring on new Awair test

balloob commented:
Great contribution, thanks! 🎉

ahayworth commented:
@zachberger I think it's lovely. 😄 If you want to help maintain this component ... that's lovely too 😉

zachberger commented:
Thanks everyone for the help getting this across the line!
